codegen/src/lib.rs

```diff
@@ -1,392 +1,17 @@
+mod lambda;
+mod main;
+
 use proc_macro::TokenStream;
-use proc_macro_error::{emit_error, proc_macro_error};
-use quote::{quote, ToTokens};
-use syn::{
-    parse_macro_input, parse_quote, spanned::Spanned, Attribute, FnArg, Ident, ItemFn, Pat, Path,
-    ReturnType, Signature, Stmt, Type,
-};
+use proc_macro_error::proc_macro_error;
 
 #[proc_macro_error]
 #[proc_macro_attribute]
-pub fn main(_attr: TokenStream, item: TokenStream) -> TokenStream {
-    let mut fn_decl = parse_macro_input!(item as ItemFn);
-
-    let wrapper = Wrapper::from_item_fn(&mut fn_decl);
-
-    let expanded = quote! {
-        #wrapper
-
-        fn __binder(
-            service: Box<dyn shuttle_service::Service>,
-            addr: std::net::SocketAddr,
-            runtime: &shuttle_service::Runtime,
-        ) -> shuttle_service::ServeHandle {
-            runtime.spawn(async move { service.bind(addr).await })
-        }
-
-        #fn_decl
-
-        #[no_mangle]
-        pub extern "C" fn _create_service() -> *mut shuttle_service::Bootstrapper {
-            let builder: shuttle_service::StateBuilder<Box<dyn shuttle_service::Service>> =
-                |factory, runtime, logger| Box::pin(__shuttle_wrapper(factory, runtime, logger));
-
-            let bootstrapper = shuttle_service::Bootstrapper::new(
-                builder,
-                __binder,
-                shuttle_service::Runtime::new().unwrap(),
-            );
-
-            let boxed = Box::new(bootstrapper);
-            Box::into_raw(boxed)
-        }
-    };
-
-    expanded.into()
-}
-
-struct Wrapper {
-    fn_ident: Ident,
-    fn_inputs: Vec<Input>,
-}
-
-#[derive(Debug, PartialEq)]
-struct Input {
-    /// The identifier for a resource input
-    ident: Ident,
-
-    /// The shuttle_service path to the builder for this resource
-    builder: Path,
+pub fn main(attr: TokenStream, item: TokenStream) -> TokenStream {
+    main::r#impl(attr, item)
```

Comment:
@NodarD  In order to keep the implementation for the `main` and `lambda` macros separate, I moved the implementation for `main` into its own file (the actual proc_macro function must remain in the lib root). The implementation for `lambda` is a `todo`, but if we want to merge this refactor, I can remove it from this branch and just keep the moved implementation of `main`

Reply:
I'm good with merging this refactor first :smile: